components/eamxx/src/physics/nudging/atmosphere_nudging.cpp

```diff
@@ -0,0 +1,264 @@
+  // Layout for horiz_wind field
+  FieldLayout horiz_wind_layout { {COL,CMP,LEV}, {m_num_cols,2,m_num_levs} };
+
+  constexpr int ps = 1;
+  auto Q = kg/kg;
+  Q.set_string("kg/kg");
+  add_field<Required>("p_mid", scalar3d_layout_mid, Pa, grid_name, ps);
+  add_field<Updated>("T_mid", scalar3d_layout_mid, K, grid_name, ps);
+  add_field<Updated>("qv", scalar3d_layout_mid, Q, grid_name, "tracers", ps);
+  add_field<Updated>("horiz_winds",   horiz_wind_layout,   m/s,     grid_name, ps);
+  
+  //Now need to read in the file
+  scorpio::register_file(datafile,scorpio::Read);
+  m_num_src_levs = scorpio::get_dimlen_c2f(datafile.c_str(),"lev");
+  scorpio::eam_pio_closefile(datafile);
+}
+
+// =========================================================================================
+void NUDGING::initialize_impl (const RunType /* run_type */)
+{
+  using namespace ShortFieldTagsNames;
+  FieldLayout scalar3d_layout_mid { {COL,LEV}, {m_num_cols, m_num_src_levs} };
+  FieldLayout horiz_wind_layout { {COL,CMP,LEV}, {m_num_cols,2,m_num_src_levs} };
+  fields_ext["T_mid"] = view_2d<Real>("T_mid",m_num_cols,m_num_src_levs);
+  auto T_mid_h       = Kokkos::create_mirror_view(fields_ext["T_mid"]);
+  host_views["T_mid"] = view_1d_host<Real>(T_mid_h.data(),T_mid_h.size());
+  layouts.emplace("T_mid", scalar3d_layout_mid);
+
+  fields_ext["p_mid"] = view_2d<Real>("p_mid",m_num_cols,m_num_src_levs);
+  auto p_mid_h       = Kokkos::create_mirror_view(fields_ext["p_mid"]);
+  host_views["p_mid"] = view_1d_host<Real>(p_mid_h.data(),p_mid_h.size());
+  layouts.emplace("p_mid", scalar3d_layout_mid);
+  
+  fields_ext["qv"] = view_2d<Real>("qv",m_num_cols,m_num_src_levs);
+  auto qv_h       = Kokkos::create_mirror_view(fields_ext["qv"]);
+  host_views["qv"] = view_1d_host<Real>(qv_h.data(),qv_h.size());
+  layouts.emplace("qv", scalar3d_layout_mid);
+
+  fields_ext_3d["horiz_winds"] = view_3d<Real>("horiz_winds",m_num_cols,2,m_num_src_levs);
+  auto horiz_winds_h       = Kokkos::create_mirror_view(fields_ext_3d["horiz_winds"]);
+  host_views["horiz_winds"] = view_1d_host<Real>(horiz_winds_h.data(),horiz_winds_h.size());
+  layouts.emplace("horiz_winds", horiz_wind_layout);
+  
+  auto grid_l = m_grid->clone("Point Grid", false);
+  grid_l->reset_num_vertical_lev(m_num_src_levs);
+  ekat::ParameterList data_in_params;
+  data_in_params.set("Field Names",m_fnames);
+  data_in_params.set("Filename",datafile);
+  data_in_params.set("Skip_Grid_Checks",true);  // We need to skip grid checks because multiple ranks may want the same column of source data.
+  AtmosphereInput data_input0(m_comm,data_in_params);
+  data_input0.init(grid_l,host_views,layouts);
+  data_input=data_input0;
+
+  T_mid_ext = fields_ext["T_mid"];
+  p_mid_ext = fields_ext["p_mid"];
+  horiz_winds_ext = fields_ext_3d["horiz_winds"];
+  qv_ext = fields_ext["qv"];
+  ts0=timestamp();
+}
+
+void NUDGING::time_interpolation (const int time_s) {
```

Comment:
It seems that this fcn runs at every time step. I'm assuming the input data is on a coarser time grid than that where the simulation runs, so this ends up re-reading the same data for a few time steps. I/O is not cheap, so we should probably store X_bef and X_aft as class members, and update them only when time_index changes.

In fact, I think this class should share a lot of code with SPA. Perhaps we should factor out some of that code? @AaronDonahue

Reply:
Thanks for the comment, I will talk to Aaron about this.

Reply:
Talking to Aaron, I think that this would be an end goal but not for this PR. It would take a bit of work to share the code and we want to get something working to provide some collaborators with nudged data. For now I will create a struct that holds the data. I think there would be a nudging 2.0 at some point that would share code with SPA and then also possibly not hard code the parameters you want to nudge but have the user be able to specify that. But we think these updates are better to hold off on for now. thanks

Reply:
ok